File: Design-Patterns-Using-JavaScript/ArchitecturalPatterns/PipeAndFilter.js
// Filter Base Class
class Filter {
    process(data) {
        throw new Error('You have to implement the method process!');
    }
}

// Filters
class CapitalizeFilter extends Filter {
    process(data) {
        return data.toUpperCase();
    }
}

class ReplaceSpaceFilter extends Filter {
    process(data) {
        return data.replace(/ /g, '_');
    }
}

class RemoveSpecialCharactersFilter extends Filter {
    process(data) {
        const specialCharacters = ",@!";
        let result = '';
        for (let i = 0; i < data.length; i++) {
            if (specialCharacters.indexOf(data[i]) === -1) {
                result += data[i];
            }
        }
        return result;
    }
}

// Data Processing Pipeline
class DataProcessingPipeline {
    constructor() {
        this.filters = [];
    }

    addFilter(filter) {
        this.filters.push(filter);
    }

    processData(data) {
        let processedData = data;
        this.filters.forEach(filter => {
            processedData = filter.process(processedData);
        });
        return processedData;
    }
}

// Client code
const pipeline = new DataProcessingPipeline();
pipeline.addFilter(new CapitalizeFilter());
pipeline.addFilter(new ReplaceSpaceFilter());
pipeline.addFilter(new RemoveSpecialCharactersFilter());

const data = "Hello, World!";
const result = pipeline.processData(data);
console.log("Result: " + result);  

File: Design-Patterns-Using-Ruby/ArchitecturalPatterns/PipeAndFilter.rb
# Filter Base Class
class Filter
    def process(data)
        raise NotImplementedError, 'Subclasses must implement the process method'
    end
end

# Filters
class CapitalizeFilter < Filter
    def process(data)
        data.upcase
    end
end

class ReplaceSpaceFilter < Filter
    def process(data)
        data.gsub(' ', '_')
    end
end

class RemoveSpecialCharactersFilter < Filter
    def process(data)
        data.gsub(/[, @!]/, '')
    end
end

# Data Processing Pipeline
class DataProcessingPipeline
    def initialize
        @filters = []
    end

    def add_filter(filter)
        @filters << filter
    end

    def process_data(data)
        @filters.inject(data) { |processed_data, filter| filter.process(processed_data) }
    end
end

# Main class
pipeline = DataProcessingPipeline.new
pipeline.add_filter(CapitalizeFilter.new)
pipeline.add_filter(ReplaceSpaceFilter.new)
pipeline.add_filter(RemoveSpecialCharactersFilter.new)

data = 'Hello, World!'
result = pipeline.process_data(data)
puts "Result: #{result}"

=begin 
Result: HELLO_WORLD
 =end